hal/ia32/memory.c

```diff
@@ -195,28 +195,69 @@ int hal_memoryGetNextEntry(addr_t start, addr_t end, mapent_t *entry)
 
 
 /* Flushes 8042 keyboard controller buffers */
-static void memory_empty8042(void)
+static int memory_empty8042(void)
 {
 	unsigned char status;
-
-	do {
+	unsigned short int timeout;
+	for (timeout = 0xFFFF; timeout > 0; --timeout) {
 		/* Discard input data */
-		if ((status = hal_inb((void *)0x64)) & 0x01) {
+		status = hal_inb((void *)0x64);
+		if ((status & 0x01) != 0) {
 			hal_inb((void *)0x60);
 			continue;
 		}
-	} while (status & 0x02);
+		if ((status & 0x02) != 0) {
+			return 0;
+		}
+	}
+	return -1;
+}
+
+
+static int checkA20Line(void)
+{
+	int result = 0;
+	/* Set base to some address in the first 1MiB that can be overwritten */
+	void *const base = (void *)0x7DFE;
+	/* Some values that should be equal, if A20 is off */
+	unsigned short int v1 = *(unsigned short int *)(base + 0x100000);
+	unsigned short int v2 = *(unsigned short int *)base;
+	if (v1 == v2) {
+		/* They are equal: try to overwrite one of them */
+		*(int *)base = v2 + 1;
```

Comment:
I forgot to change it, while I was changing `int` to `short int`. I've also just realized that `u16` is defined there, so I'll use that instead of `unsigned short int`.